lib/transformer.cc

```diff
@@ -0,0 +1,176 @@
+// -*- mode: c++; c-basic-offset: 2; indent-tabs-mode: nil; -*-
+// Copyright (C) 2014 Henner Zeller <h.zeller@acm.org>
+// Copyright (C) 2015 Christoph Friedrich <christoph.friedrich@vonaffenfels.de>
+//
+// This program is free software; you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation version 2.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://gnu.org/licenses/gpl-2.0.txt>
+
+#include <assert.h>
+
+#include "transformer.h"
+
+namespace rgb_matrix {
+
+RotateCanvas::RotateCanvas(int angle) : delegatee_(NULL), angle_(angle) {
+  // Only allow angles in 90° steps
+  assert(angle % 90 == 0);
+}
+
+void RotateCanvas::SetDelegatee(Canvas* delegatee) {
+  assert(delegatee != NULL);
+
+  pivot_x_ = (delegatee->width() - 1) / 2;
+  pivot_y_ = (delegatee->height() - 1) / 2;
+  delegatee_ = delegatee;
+}
+
+void RotateCanvas::SetPixel(int x, int y, uint8_t red, uint8_t green, uint8_t blue) {
+  assert(delegatee_ != NULL);
+
+  const float s = (angle_ == 90 ? 1 : (angle_ == 270 ? -1 : 0));
+  const float c = (angle_ == 180 ? -1 : (angle_ == 0 ? 1 : 0));
+
+  // Offset needed cause of little precision errors on 180° and 270°
+  const int offset_x = (angle_ == 90 || angle_ == 180 ? 1 : 0);
+  const int offset_y = (angle_ == 180 || angle_ == 270 ? 1 : 0);
+
+  // translate point to origin
+  x -= pivot_x_;
+  y -= pivot_y_;
+
+  float rot_x = x * c - y * s;
+  float rot_y = x * s + y * c;
+
+  // translate back
+  x = rot_x + pivot_x_ + offset_x;
+  y = rot_y + pivot_y_ + offset_y;
+
+  delegatee_->SetPixel(x, y, red, green, blue);
+}
+
+int RotateCanvas::width() const { 
+  assert(delegatee_ != NULL);
```

Comment:
So delete the assertion and with a private RotateCanvas we should always think there is a canvas there. So only test against NULL in the Transform() function.